<launch>

	<!-- Base Parameters -->
	<arg name="env_id" default="17DRP5sb8fy" />
	<arg name="data_path" default="data_litevloc/vnav_eval/matterport3d" />
	<arg name="map_path" default="$(arg data_path)/s$(arg env_id)/merge_finalmap" />
	<arg name="setup_simu_env" default="true" />
	<arg name="use_rviz" default="true" />
	<arg name="use_nav" default="true" />
	<arg name="gazebo_gui" default="false" />
	<param name="use_sim_time" value="true" />

	<!-- Localization Parameters -->
	<!-- trigger global localization if distance to the closest map node exceeds it -->
	<!-- indoor: 10.0; outdoor: 15.0 -->
	<arg name="global_pos_threshold" default="10.0" /> 
	<arg name="vloc_freq" default="1" />

	<!-- Launch Environment -->
	<group if="$(arg setup_simu_env)">
		<arg name="useLocalPlanner" default="false" />
		<include file="$(find benchmark_visual_nav)/environment/launch/matterport/system_$(arg env_id).launch" >
			<arg name="project_directory" value="$(find benchmark_visual_nav)" />
			<arg name="environment_directory" value="$(find benchmark_visual_nav)/data/matterport/$(arg env_id)" />
			<arg name="gazebo_gui" value="$(arg gazebo_gui)" />
			<arg name="useLocalPlanner" value="$(arg useLocalPlanner)" />
			<arg name="use_rviz" value="false" />
			
			<group if="$(arg useLocalPlanner)">
				<arg name="odomTopic" value="/state_estimation" />
			</group>
			<group unless="$(arg useLocalPlanner)">
				<arg name="odomTopic" value="/vloc/state_estimation" />
			</group>
		</include>
	</group>

	<!-- Navigation Interface -->
	<!-- Transform the frame_id of localization odometry to robot -->
	<!-- Publish: /vloc/state_estimation, /vloc/state_estimation_path -->
	<group if="$(arg use_nav)">
		<node pkg="navigation_interface" type="vis_navigation_interface" name="vis_navigation_interface" output="log">
			<param name="loc_world_frame_id" value="vloc_map" />
			<param name="loc_sensor_frame_id" value="habitat_camera" />
			<param name="world_frame_id" value="vloc_map_not_used" />
			<param name="sensor_frame_id" value="habitat_camera" />
			<param name="base_frame_id" value="vehicle" /> <!-- robot_id -->
			<param name="point_skip" type="int" value="5" />

			<remap from="/odometry" to="/pose_fusion/odometry" />
			<remap from="/camera_info" to="/habitat_camera/color/camera_info" />
			<remap from="/depth_image" to="/habitat_camera/depth/image" />
		</node>	
	</group>

	<!-- Planning -->
	<group if="$(arg setup_simu_env)">
		<node name="ros_global_planner" pkg="litevloc" type="ros_global_planner.py" output="log"
			args="--map_path $(arg map_path) \
						--image_size 512 288 --device cuda \
						--vpr_method cosplace --vpr_backbone ResNet18 --vpr_descriptors_dimension 256" >
			<param name="frame_id_map" value="vloc_map" />
			<param name="main_freq" value="1" />
			<param name="conv_dist" value="0.5" />
		</node>

		<node name="ros_publish_goal_image" pkg="litevloc" type="ros_publish_goal_image.py" output="log"
			args="--map_path $(arg map_path)" >
		</node>
	</group>

	<!-- Localization -->
	<node name="ros_loc_pipeline" pkg="litevloc" type="ros_loc_pipeline.py" output="screen"
		args="--map_path $(arg map_path)
					--image_size 512 288
					--device cuda --vpr_method cosplace --vpr_backbone ResNet18 --vpr_descriptors_dimension 256 
					--img_matcher master --pose_solver pnp 
					--config_pose_solver $(find litevloc)/python/config/dataset/matterport3d.yaml
					--ros_rgb_img_type raw
					--global_pos_threshold $(arg global_pos_threshold)
					--min_master_conf_thre 1.5
					--min_kpts_inliers_thre 200 
					--min_solver_inliers_thre 200 --viz" >
		<remap from="/color/image" to="/habitat_camera/color/image" />
		<remap from="/depth/image" to="/habitat_camera/depth/image" />
		<remap from="/color/camera_info" to="/habitat_camera/color/camera_info" />
		<param name="frame_id_map" value="vloc_map" />
		<param name="main_freq" value="$(arg vloc_freq)" />
		<param name="min_depth" value="0.001" />
		<param name="max_depth" value="10.0" />
	</node>

	<node name="depth_registration" pkg="litevloc" type="depth_registration.py" output="log">
		<remap from="/depth/image" to="/habitat_camera/depth/image" />
		<remap from="/depth/camera_info" to="/habitat_camera/depth/camera_info" />
		<param name="frame_id_map" value="vloc_map" />
		<param name="voxel_radius" value="0.1" />
		<param name="min_depth" value="0.001" />
		<param name="max_depth" value="10.0" />
	</node>

	<node name="ros_pose_fusion" pkg="litevloc" type="ros_pose_fusion.py" output="log"
		args="--odom_type depth_reg">
		<remap from="/local/odometry" to="/depth_reg/odometry" />
		<remap from="/global/odometry" to="/vloc/odometry" />
		<param name="frame_id_map" value="vloc_map" />
	</node>

  <!-- Static TF -->
  <!-- Option 1: x y z yaw pitch roll (radians) frame_id child_frame_id period_in_ms -->
  <!-- Option 2: x y z qx qy qz qw frame_id child_frame_id period_in_ms -->
  <!-- <node pkg="tf" type="static_transform_publisher" name="vloc_trans_publisher" args="0 0 0 0 0 0 map vloc_map 10" /> -->
	
	<!-- RVIZ -->
	<group if="$(arg use_rviz)">
		<node name="rviz" pkg="rviz" type="rviz" args="-d $(find litevloc)/rviz_cfg/topo_graph_matterport3d.rviz" required="false" />
	</group>

</launch>
